### PR TITLE
New SMS Listeners for 2.32

### PR DESCRIPTION
**Pull request by JasperTimm**
This PR is to merge the new SMS handling which is being used to send SMS via the new Android Capture App.

It uses a shared SMS compression library which is currently hosted in the eHA org, but should be migrated to DHIS2: https://github.com/eHealthAfrica/dhis2-sms-compression

This PR is for a special version of 2.32 with SMS that can be used by early adopters to test before 2.33 is released. Another PR will be created to merge into master/2.33.

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/AggregateDatasetSMSListener.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+import org.hisp.dhis.sms.incoming.IncomingSms;
+import org.hisp.dhis.smscompression.SMSConsts.SubmissionType;
+import org.hisp.dhis.smscompression.SMSResponse;
+import org.hisp.dhis.smscompression.models.AggregateDatasetSMSSubmission;
+import org.hisp.dhis.smscompression.models.SMSDataValue;
+import org.hisp.dhis.smscompression.models.SMSSubmission;
+import org.hisp.dhis.smscompression.models.UID;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.user.UserService;
+import org.jfree.util.Log;
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/AggregateDatasetSMSListener.java]: Can we use LogFactory from Log4J here, as we use that across the system?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/AggregateDatasetSMSListener.java [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        }
+
+        CategoryOptionCombo aoc = categoryService.getCategoryOptionCombo( aocid.uid );
+        if ( aoc == null )
+        {
+            throw new SMSProcessingException( SMSResponse.INVALID_AOC.set( aocid ) );
+        }
+
+        // TODO: This seems a bit backwards, why is there no
+        // dataSet.hasOrganisationUnit?
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/AggregateDatasetSMSListener.java]: You could add that method on DataSet then.

File: dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/system/util/SmsUtils.java [hunk 3]
```diff
@@ -73,6 +73,24 @@ public static String getCommandString( String text )
 
         return commandString;
     }
+    
+    public static boolean isBase64( IncomingSms sms ) {
```
Comment (larshelge) [inline dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/system/util/SmsUtils.java]: Please apply the DHIS 2 code style to this class.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/NewSMSListener.java [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+    {
+        List<Program> programs = programService.getAllPrograms();
+
+        return programs.stream().map( o -> getIdFromMetadata( o, lastSyncDate ) ).filter( Objects::nonNull )
+            .collect( Collectors.toList() );
+    }
+
+    private List<SMSMetadata.ID> getAllOrgUnitIds( Date lastSyncDate )
+    {
+        List<OrganisationUnit> orgUnits = organisationUnitService.getAllOrganisationUnits();
```
Comment (zubaira) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/NewSMSListener.java]: Getting all OrganisationUnits is quite a heavy operation. We can move the logic to SQL level in order to fetch only those which are needed.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/NewSMSListener.java [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+
+        return orgUnits.stream().map( o -> getIdFromMetadata( o, lastSyncDate ) ).filter( Objects::nonNull )
+            .collect( Collectors.toList() );
+    }
+
+    private List<SMSMetadata.ID> getAllDataElements( Date lastSyncDate )
+    {
+        List<DataElement> dataElements = dataElementService.getAllDataElements();
```
Comment (zubaira) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/NewSMSListener.java]: same as OUs.

**Comment by larshelge** [comment]
I am closing this PR for now. Please target the master branch first, and then we can back-port the commit later. Please look into the suggested fixes to master PR.

**Pull request closed without merging**